service/src/main/java/bio/terra/cbas/controllers/util/AsyncExceptionHandler.java

```diff
@@ -0,0 +1,21 @@
+package bio.terra.cbas.controllers.util;
+
+import java.lang.reflect.Method;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.aop.interceptor.AsyncUncaughtExceptionHandler;
+
+public class AsyncExceptionHandler implements AsyncUncaughtExceptionHandler {
+
+  private final Logger logger = LoggerFactory.getLogger(AsyncExceptionHandler.class);
+
+  @Override
+  public void handleUncaughtException(Throwable ex, Method method, Object... params) {
```

Comment:
`params` contain the parameters that were passed to the async method, here `triggerWorkflowSubmission`. I had attempted to extract RunSet ID from the `params` but it wasn't trivial. I can try again.

>And should we be updating the database with failure states when a run creation has an error?

I had thought about that, but I wasn't sure how to have `RunSetDao` and `RunDao` references in the `AsyncExceptionHandler` class. One way would be to have RunSetApiController itself pass runSetDao and runDao reference to triggerWorkflowSubmission so that it can be extracted from params but I am not sure if that is the right approach.

Reply:
> One way would be to have RunSetApiController itself pass runSetDao and runDao reference to triggerWorkflowSubmission so that it can be extracted from params but I am not sure if that is the right approach.

I don't have a better suggestion - do you have a specific concern about doing it this way?

(FWIW, it's not obvious to me why you need the references to extract them from params. I'm sure the devil is in the details...)

Reply:
Another way would be to make AsyncExceptionHandler a component and have RunSetDao and RunDao references be auto-wired 🤔

Reply:
In [this commit](https://github.com/DataBiosphere/cbas/pull/281/commits/6ef01a8d2a42f212184331380022f159b3f067aa) I extracted the Run Set ID that is available in params and added it to the log message

Reply:
> One way would be to have RunSetApiController itself pass runSetDao and runDao reference to triggerWorkflowSubmission so that it can be extracted from params but I am not sure if that is the right approach.

I actually take back this suggestion. Apart from `triggerWorkflowSubmission` there are other methods in the RunSetService which do use RunSetDao and RunDao. Implementing this suggestion would mean each method needs to have these references passed in. This would also mean that RunSetsApiController needs to have RunDao as its member again (this was removed in this PR as some DAO logic was moved into RunSetService)

Reply:
It feels weird to me to add DAOs to method signatures just so that an exception handler can extract them again from its `params` object.

It feels must more natural to me for the exception handler to have references to the DAO if part of exception handling is updating database tables.